audio/src/soundio.zig

```diff
@@ -0,0 +1,201 @@
+const std = @import("std");
+const mem = std.mem;
+const fmt = std.fmt;
+const c = @import("soundio").c;
+const SoundIo = @import("soundio").SoundIo;
+const SoundIoDevice = @import("soundio").Device;
+const SoundIoInStream = @import("soundio").InStream;
+const SoundIoOutStream = @import("soundio").OutStream;
+const Mode = @import("main.zig").Mode;
+const DeviceDescriptor = @import("main.zig").DeviceDescriptor;
+
+const is_raw_sign = "MD_IS_RAW";
+const input_mode_sign = "MD_INPUT_";
+const output_mode_sign = "MD_OUTPUT";
+
+const Audio = @This();
+
+const DeviceUnion = union(Mode) {
+    input: SoundIoInStream,
+    output: SoundIoOutStream,
+};
+
+pub const DeviceDescriptorInternal = SoundIoDevice;
+pub const Device = struct {
+    handle: DeviceUnion,
+
+    pub fn deinit(self: Device) void {
+        return switch (self.handle) {
+            .input => |d| d.deinit(),
+            .output => |d| d.deinit(),
+        };
+    }
+};
+pub const DeviceIterator = struct {
+    ctx: Audio,
+    mode: Mode,
+    device_len: u16,
+    index: u16,
+
+    pub fn next(self: *DeviceIterator) IteratorError!?DeviceDescriptor {
+        if (self.index < self.device_len) {
+            const device_desc = switch (self.mode) {
+                .input => self.ctx.handle.getInputDevice(self.index) orelse return null,
+                .output => self.ctx.handle.getOutputDevice(self.index) orelse return null,
+            };
+            self.index += 1;
+            return DeviceDescriptor{
+                .id = try fmt.allocPrint(
+                    self.ctx.allocator,
+                    "{s}[{s}][{s}]",
+                    .{
+                        device_desc.id(),
+                        is_raw_sign,
+                        switch (@intToEnum(Mode, device_desc.handle.aim)) {
+                            .input => input_mode_sign,
+                            .output => output_mode_sign,
+                        },
+                    },
```

Comment:
Why are we storing so much information with id. DeviceDescriptor already has is_raw and mode fields.

Reply:
so if `internal` and `parseID` get removed, which is correct strategy?
- `InvalidParams` error, if `id` exist but `mode` and `is_raw` was **null**
- automatically connect to best device, if `id` exist but `mode` and `is_raw` was **null**

Reply:
I think we had discussed that we may want to allow device creation just from id, and if I understand it correctly id doesn't have enough information to uniquely identify the devices. So that does justifies what you did above. But it has its problems. 

The id held by soundio we actually get it from the internal audio server itself, so if we made id a custom format, it would be difficult to use the id in future non-soundio native backends. 

We can in theory make the id format a standard which all backends must adhere to but it has the drawback that id would have to be allocated. Alternatively we can just not allow creating device solely from id but require the complete descriptor. cc: @slimsag

Reply:
> The id held by soundio we actually get it from the internal audio server itself, so if we made id a custom format, it would be difficult to use the id in future non-soundio native backends.

I'm not worried about this. In my view, ID should be clearly marked as "a string of garbage only this backend understands" - nobody except that backend should try to interpret it, and there could be changes to the backend (like if we ever move away from soundio) that invalidate those IDs. It's safe to store it in a file and get the device from that ID, but only using the same backend - and not if the backend changes in the future.